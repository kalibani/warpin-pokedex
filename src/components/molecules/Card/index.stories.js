// Stories Card Component
// --------------------------------------------------------
import React, { Fragment } from 'react';
import { storiesOf } from '@storybook/react';
import Card from './index';
import Notes from './readme.md';

storiesOf('Atoms', module).add('Card', () => (
  <Card />
), {
  info: { inline: true, header: false, text: Notes }
});
