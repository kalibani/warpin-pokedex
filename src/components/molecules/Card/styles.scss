/* ==========================================================================
Styles for Card component
========================================================================== */

// import variables
@import "../../../assets/styles/variables";


// Start styles Card component
.m-card {
  padding: 15px;
  margin-bottom: -20px;

  .card-clickable {
    background: #49505733;
    float: right;
    color: #f8f9fa;
    border-radius: 50%;
    padding: 5px;
    position: relative;
    z-index: 10;
    font-size: 12px;
    top: 50px;
    left: -32px;
    cursor: pointer;
    &:hover {
      background: rgba(0,123,255,.5);
    }
  }

  .card-clickable-area {
    display: block;
    position: absolute;
    z-index: 15;
    background: $white;
    color: $outer-space;
    padding: 10px;
    width: 15%;
    border-radius: 7px;
    font-size: 13px;
    margin-top: 83px;
    margin-left: 10px;
    box-shadow: 20px;
    svg {
      float: left;
      position: relative;
      left: 126px;
      top: -25px;
      color: $white;
      font-size: 23px;
    }
    .text-capitalize {
        &:last-child {
          span {

            display: none;
          }
        }
    }
  }

  .card-image-wrapper {
    border-radius: 10px;
    min-height: calc(180px * 1.5);
    width: 90%;
    background: linear-gradient(to right, rgba(3, 19, 65, 0.52) 0%, rgba(38, 3, 65, 0.08) 100%);

    overflow: hidden;
    display: flex;
    justify-content: center;
    align-items: center;
    box-shadow: 1px 2px 4px 2px rgba(0, 0, 0, .5);
    &.animation {
      animation: pulse 5s infinite;
    }
    &:hover {
      cursor: pointer;
      transition: 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275) transform;
      transform: translateY(-10px);
      box-shadow: 1px 2px 4px 2px rgba(0, 0, 0, .5);
      animation: stretch
      1.5s
      ease-out
      0s
      alternate
      infinite
      none
      running;
    }
    .image-card {
      width: 100%;
      height: 100%;
    }

    @keyframes pulse {
      0% {
        background-color: $gradient-grey;
      }
      100% {
        background-color: $gradient-blue;
      }
    }

    @keyframes stretch {
      
      50% {
        background-color: $gradient-grey;
      }
      100% {
        transform: scale(1);
        background-color: $gradient-blue;
      }
    }
  }
}
