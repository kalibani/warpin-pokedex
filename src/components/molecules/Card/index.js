/* eslint-disable import/no-cycle */
// Card Component
// --------------------------------------------------------

import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classname from 'classnames';
import { H4 } from 'components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faEllipsisH, faCaretUp } from '@fortawesome/free-solid-svg-icons';
import './styles.scss';

const Card = ({
  srcSet, src, className, name, height, weight, types, abilities, statistics
}) => {
  const [isShow, setIsShow] = useState(false);
  const handleClickShow = (payload) => {
    setIsShow(payload);
  };
  useEffect(() => {
    window.onmousedown = () => {
      handleClickShow(false);
    };
  });
  const animatiom = !src ? 'animation' : '';
  const classNames = classname('m-card', className);
  return (
    <div className={classNames}>
      <div className="card-clickable" role="presentation" onClick={() => handleClickShow(true)}>
        <FontAwesomeIcon icon={faEllipsisH} />
      </div>
      {
        isShow && (
        <div className="card-clickable-area">
          <FontAwesomeIcon icon={faCaretUp} />
          <H4 className="text-dark text-capitalize">{name}</H4>
          <div className="card-info">
            <div className="card-info-wrapper d-flex justify-content-between">
              <div>Height </div>
              <div>{height}</div>
            </div>
            <div className="card-info-wrapper d-flex justify-content-between">
              <div>Weight </div>
              <div>{weight}</div>
            </div>
            <div className="card-info-wrapper d-flex justify-content-between">
              <div> Types </div>
              <div>
                {
                      types.map((element, index) => (
                        <div className="text-capitalize" key={index}>
                          { element.type.name}
                          <span>,</span>
                        </div>
                      ))
                    }
              </div>
            </div>
            <div className="card-info-wrapper d-flex justify-content-between">
              <div> Abilities </div>
              <div>
                {
                      abilities.map((element, index) => (
                        <div className="text-capitalize" key={index}>
                          { element.ability.name}
                          <span>,</span>
                        </div>
                      ))
                    }
              </div>
            </div>
            <div className="card-info-wrapper">
              <div> Stats :</div>
              {
                    statistics.map((element, index) => (
                      <div className="d-flex justify-content-between" key={index}>
                        <div className="text-capitalize">
                          { element.stat.name}
                        </div>
                        <div className="text-capitalize">
                          { element.base_stat}
                        </div>
                      </div>
                    ))
                  }
            </div>
          </div>
        </div>
        )
      }
      <div className={`card-image-wrapper ${animatiom}`}>
        {
          src ? (
            <img
              className="lazy image-card"
              srcSet={srcSet}
              src={src}
              alt="card-images"
            />
          ) : (
            <div>
              No Image Availabe
            </div>
          )
        }
      </div>
      <H4 className="text-dark text-capitalize ml-2 mt-3">
        {name}
      </H4>
    </div>

  );
};

Card.propTypes = {
  src: PropTypes.string,
  srcSet: PropTypes.string,
  className: PropTypes.string,
  name: PropTypes.string,
  height: PropTypes.number,
  weight: PropTypes.number,
  types: PropTypes.array,
  abilities: PropTypes.array,
  statistics: PropTypes.array
};

Card.defaultProps = {
  src: '',
  srcSet: '',
  className: '',
  name: '',
  height: 0,
  weight: 0,
  types: [],
  abilities: [],
  statistics: []
};

export default Card;
