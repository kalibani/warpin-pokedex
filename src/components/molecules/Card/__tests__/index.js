// Test Card Component
// --------------------------------------------------------

import React from 'react';
import Card from '../index';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faEllipsisH } from '@fortawesome/free-solid-svg-icons';
const shallowSetup = () => {
  // Sample props to pass to our shallow render
  const props = {
    srcSet: 'https://dummy-url',
    src:'https://dummy-url',
    className: '',
    name: '',
    height: 0,
    weight: 0,
    types: [],
    abilities: [],
    statistics: [],
    isShow: true
  }

  const mockOnClick = jest.fn();
  // wrapper instance around rendered output
  const enzymeWrapper = shallow(<Card {...props} onClick={mockOnClick} />);

  return {
    props,
    enzymeWrapper,
    mockOnClick
  };
}

describe('<Card />', () => {
  it('should render a card with the props', () => {
    // Setup wrapper and assign props.
    const { enzymeWrapper, props, mockOnClick } = shallowSetup();
    
    expect(enzymeWrapper.find('img').hasClass('lazy image-card')).toBe(true);
    enzymeWrapper.find('div.card-clickable').simulate('click');
    // expect(mockOnClick.mock.calle.length).toEqual(1);
    expect(enzymeWrapper.find('img').at(0).props().srcSet).toBe(props.srcSet);
    expect(enzymeWrapper.containsMatchingElement(<FontAwesomeIcon icon={faEllipsisH} />)).toBe(true);
    expect(toJson(enzymeWrapper)).toMatchSnapshot();
  });
});


