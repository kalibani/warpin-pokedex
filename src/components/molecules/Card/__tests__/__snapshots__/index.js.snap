// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render a card with the props 1`] = `
<div
  className="m-card"
>
  <div
    className="card-clickable"
    onClick={[Function]}
    role="presentation"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f141",
            "M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z",
          ],
          "iconName": "ellipsis-h",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
  <div
    className="card-clickable-area"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            320,
            512,
            Array [],
            "f0d8",
            "M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z",
          ],
          "iconName": "caret-up",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <Memo(H4)
      className="text-dark text-capitalize"
    />
    <div
      className="card-info"
    >
      <div
        className="card-info-wrapper d-flex justify-content-between"
      >
        <div>
          Height 
        </div>
        <div>
          0
        </div>
      </div>
      <div
        className="card-info-wrapper d-flex justify-content-between"
      >
        <div>
          Weight 
        </div>
        <div>
          0
        </div>
      </div>
      <div
        className="card-info-wrapper d-flex justify-content-between"
      >
        <div>
           Types 
        </div>
        <div />
      </div>
      <div
        className="card-info-wrapper d-flex justify-content-between"
      >
        <div>
           Abilities 
        </div>
        <div />
      </div>
      <div
        className="card-info-wrapper"
      >
        <div>
           Stats :
        </div>
      </div>
    </div>
  </div>
  <div
    className="card-image-wrapper "
  >
    <img
      alt="card-images"
      className="lazy image-card"
      src="https://dummy-url"
      srcSet="https://dummy-url"
    />
  </div>
  <Memo(H4)
    className="text-dark text-capitalize ml-2 mt-3"
  />
</div>
`;
